### PR TITLE
Add Metrics + Metrics server

### PR DESCRIPTION
**Pull request by astoycos**
Starts to fix https://github.com/kubernetes-sigs/kpng/issues/351

### CODE REVIEW

File: hack/test_e2e.sh [hunk 1]
```diff
@@ -1147,8 +1157,9 @@ suffix=""
 cluster_count="1"
 erase_clusters=false
 print_report=false
+run_tests_on_existing_cluster=false
 
-while getopts "i:b:B:cdD:eE:n:ps:" flag
+while getopts "i:b:B:cdD:eE:n:ps:t" flag
```
Comment (jayunit100) [inline hack/test_e2e.sh:1162]:
```suggestion
while getopts "i:b:B:cdD:eE:n:ps:t:m" flag
```

File: cmd/go.mod [hunk 2]
```diff
@@ -14,30 +14,62 @@ require (
 	google.golang.org/protobuf v1.28.1
 	k8s.io/client-go v0.25.2
 	k8s.io/klog/v2 v2.80.1
+	sigs.k8s.io/kpng/api v0.0.0-20221014160047-e8543ceff80f
```
Comment (astoycos) [inline cmd/go.mod:18]: We shouldn't need to add these

File: cmd/go.mod [hunk 3]
```diff
@@ -14,30 +14,62 @@ require (
...
 )
 
 require (
+	github.com/Microsoft/go-winio v0.4.17 // indirect
+	github.com/Microsoft/hcsshim v0.9.4 // indirect
+	github.com/beorn7/perks v1.0.1 // indirect
+	github.com/blang/semver/v4 v4.0.0 // indirect
+	github.com/cespare/xxhash v1.1.0 // indirect
+	github.com/cespare/xxhash/v2 v2.1.2 // indirect
+	github.com/cilium/ebpf v0.8.1 // indirect
```
Comment (astoycos) [inline cmd/go.mod:37]: Yeah something is going on here we shouldn't have a dependency on cilium/ebpf here

File: hack/test_e2e.sh (hunk 1 above)
Comment (astoycos) [reply hack/test_e2e.sh]: Done

**Comment by astoycos** [comment]
I think I got to most of your recommendations here @jayunit100

File: server/jobs/api2local/api2local.go [hunk 4]
```diff
@@ -31,18 +31,18 @@ import (
...
-//     type MySink struct {
-//         api2local.Config
-//     }
-//
+//	type MySink struct {
+//	    api2local.Config
+//	}
 type Config struct {
-	NodeName string
+	NodeName      string
```
Comment (aryan9600) [inline server/jobs/api2local/api2local.go:45]:
```suggestion
	NodeName string
```

File: server/jobs/kube2store/kube2store.go [hunk 5]
```diff
@@ -34,6 +34,8 @@ import (
 )
 
 type Config struct {
+	UseSlices     bool
```
Comment (aryan9600) [inline server/jobs/kube2store/kube2store.go:37]:
```suggestion
	UseSlices bool
```

File: server/pkg/metrics/metrics.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+
+	"k8s.io/klog/v2"
+)
+
+var Kpng_k8s_api_events = prometheus.NewCounter(prometheus.CounterOpts{
```
Comment (aryan9600) [inline server/pkg/metrics/metrics.go:18]:
```suggestion
var KpngK8SApiEvents = prometheus.NewCounter(prometheus.CounterOpts{
```

File: server/pkg/metrics/metrics.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+	Name: "kpng_k8s_api_events_total",
+	Help: "The total number of received events from the Kubernetes API",
+})
+
+var Kpng_node_local_events = prometheus.NewCounter(prometheus.CounterOpts{
```
Comment (aryan9600) [inline server/pkg/metrics/metrics.go:23]:
```suggestion
var KpngNodeLocalEvents = prometheus.NewCounter(prometheus.CounterOpts{
```

File: server/jobs/kube2store/kube2store.go (hunk 5 above)
Comment (aryan9600) [reply server/jobs/kube2store/kube2store.go:37]: Also, this field doesn't seem to be used anywhere else in this PR

File: cmd/kpng/main.go [hunk 8]
```diff
@@ -23,16 +23,20 @@ import (
...
 	"k8s.io/klog/v2"
 
 	"sigs.k8s.io/kpng/server/pkg/proxy"
 )
 
 var (
-	cpuprofile = flag.String("cpuprofile", "", "write cpu profile to file")
-	version    = "(unknown)"
+	cpuprofile    = flag.String("cpuprofile", "", "write cpu profile to file")
+	exportMetrics = flag.String("exportMetrics", "", "start metrics server on the specified IP:PORT")
```
Comment (aryan9600) [inline cmd/kpng/main.go:37]:
```suggestion
	metricsServer = flag.String("metrics-server", "", "start metrics server on the specified IP:PORT")
```

**Review by aryan9600 — CHANGES_REQUESTED** [review]

File: server/pkg/metrics/metrics.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+	Name: "kpng_node_local_events_total",
+	Help: "The total number of received events from the Kubernetes API for a given node",
+})
```
Comment (aryan9600) [inline server/pkg/metrics/metrics.go:26]:
```suggestion
var KpngNodeLocalEvents = prometheus.NewCounterVec(prometheus.CounterOpts{
	Name: "kpng_node_local_events_total",
	Help: "The total number of received events from the Kubernetes API for a given node",
}, []strings{"node_name"})
```

File: server/pkg/server/watchstate/watchstate.go [hunk 10]
```diff
@@ -109,6 +110,7 @@ func (w *WatchState) send(item *localnetv1.OpItem) {
 	if w.Err != nil {
 		return
 	}
+	metrics.Kpng_node_local_events.Inc()
```
Comment (aryan9600) [inline server/pkg/server/watchstate/watchstate.go:113]: This should increment the counter per label, i.e. per node name. But I don't think we have access to the node name here, maybe we need to move this higher up?

Comment (aryan9600) [reply server/pkg/server/watchstate/watchstate.go:113]: okay that works for now ig :)

**Comment by jayunit100** [comment]
can we merge and iterate on this @aryan9600 @astoycos ?

**Comment by jayunit100** [comment]
/lgtm
/approve

lets get this in then we can build more fine grained metrics into it

**Pull request merged**